# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [1]:
import numpy as np
from scipy import stats
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

**Гіпотези**   

$H_0$: $\mu = 123.7$ (середній час до покупки)  
$H_a$: $\mu > 123.7$ (середній час підвищився)  
Тип тесту: правосторонній (one-tailed)

Ми перевіряємо, чи **збільшився** середній час до покупки після редизайну.

In [2]:
# Дані з умови
time = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7
alpha = 0.1

# Обчислимо вибіркову статистику
x_bar = np.mean(time)
s = np.std(time, ddof=1)
n = len(time)

# Обчислимо t-статистику вручну
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

# Критичне значення для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1 - alpha, df)

# p-value
p_value = 1 - stats.t.cdf(t_stat, df)

print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"p-value: {p_value:.4f}")

# Висновок
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
t-статистика: 2.069
Критичне значення (df=6): 1.440
p-value: 0.0420
Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості 0.1


In [3]:
# scipy тест
res = stats.ttest_1samp(time, popmean=mu0, alternative="greater")
print(f"scipy: t = {res.statistic:.3f}, p-value = {res.pvalue:.5f}, df = {res.df:.3f}")

# висновок на основі р-значення
if res.pvalue < alpha:
    print(f"Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

# висновок на основі t-критичного
t_crit = stats.t.ppf(1 - alpha, res.df)

if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

scipy: t = 2.069, p-value = 0.04202, df = 6.000
Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості 0.1
Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості 0.1


**Тож, відхиляємо $H_0$, тому що на рівні значущості 0.1 є статистично значущі докази, що середній time-to-purchase після редизайну збільшився порівняно з 123.7 хв.**

In [4]:
# Змінюємо рівень значущості
alpha = 0.05

# Критичне значення для правостороннього тесту
t_crit = stats.t.ppf(1 - alpha, df)

# p-value
p_value = 1 - stats.t.cdf(t_stat, df)

print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"p-value: {p_value:.4f}")

# Висновок
if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
t-статистика: 2.069
Критичне значення (df=6): 1.943
p-value: 0.0420
Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості 0.05


In [5]:
# scipy тест
res = stats.ttest_1samp(time, popmean=mu0, alternative="greater")
print(f"scipy: t = {res.statistic:.3f}, p-value = {res.pvalue:.5f}, df = {res.df:.3f}")

# висновок на основі р-значення
if res.pvalue < alpha:
    print(f"Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

# висновок на основі t-критичного
t_crit = stats.t.ppf(1 - alpha, res.df)

if t_stat > t_crit:
    print(f"Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

scipy: t = 2.069, p-value = 0.04202, df = 6.000
Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості 0.05
Відхиляємо H0: середній час до покупки вищий за 123.7 на рівні значущості 0.05


**Отже, на рівні значущості 0.05 також є статистично значущі докази, що середній time-to-purchase після редизайну збільшився порівняно з 123.7 хв. Тому, відхиляємо $H_0$**

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [6]:
count = 311
nobs = 1200
p0 = 0.23      # гіпотетична пропорція (H0)

# z-тест для пропорції
stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha=0.03, method='normal')

print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")

Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


**Висновок:**
* $H_0: p = 0.23$
* $H_а: p > 0.23$
* Оскільки $p\text{-value} = 0.0106 < 0.03$, відхиляємо $H_0$.
   * 👉 Є статистично значущі докази, що **тепер більше ніж 23%** дорослих упізнають логотип компанії.

**97% довірчий інтервал (0.232, 0.287)**
* Це означає, що з імовірністю 95% справжня частка лежить у межах **від 23.2% до 28.7%**.